pypose/module/dynamics.py

```diff
@@ -0,0 +1,196 @@
+import torch as torch
+import torch.nn as nn
+import pypose as pp
+from torch.autograd.functional import jacobian
+
+
+class _System(nn.Module):
+    def __init__(self, time=False):
+        super().__init__()
+        self.jacargs = {'vectorize':True, 'strategy':'reverse-mode'}
+        if time:
+            self.register_buffer('t',torch.zeros(1))
+            self.register_forward_hook(self.forward_hook)
+
+    def forward_hook(self, module, inputs, outputs):
+        self.input, self.state = inputs
+        self.t.add_(1)
+
+    def forward(self, state, input):
+        state = self.state_transition(state, input)
+        return self.observation(state, input)
+
+    def state_trasition(self):
+        pass
+
+    def observation(self):
+        pass
+
+    def reset(self,t=0):
+        self.t.fill_(0)
+
+    @property
+    def A(self):
+        if hasattr(self, '_A'):
+            return self._A
+        else:
+            func = lambda x: self.state_trasition(x, self.input)
+            return jacobian(func, self.state, **self.jacargs)
+
+    @property
+    def B(self):
+        if hasattr(self, '_B'):
+            return self._B
+        else:
+            func = lambda x: self.state_trasition(self.state, x)
+            return jacobian(func, self.input, **self.jacargs)
+
+    @property
+    def C(self):
+        if hasattr(self, '_C'):
+            return self._C
+        else:
+            func = lambda x: self.observation(x, self.input)
+            return jacobian(func, self.state, **self.jacargs)
+ 
+    @property
+    def D(self):
+        if hasattr(self, '_D'):
+            return self._D
+        else:
+            func = lambda x: self.observation(self.state, x)
+            return jacobian(func, self.input, **self.jacargs)
+
+
+class LTI(_System):
+    r'''
+    A sub-class of '_System' to represent Linear Time-Invariant system.
+    
+    Args:
+        A, B, C, D (:obj:`Tensor`): The coefficient matrix in the state-space equation of LTI system
+        c1, c2 (:obj:`Tensor`): Bias generated by system
+        
+    Note:
+        According to the actual physical meaning, the dimensions of A, B, C, D must be the same,
+        whether in the batch case or not.
+        
+        The system is time invariant.
+    '''
+    def __init__(self, A, B, C, D, c1=None, c2=None):
+        super(LTI, self).__init__(time=False)
+        assert A.ndim == B.ndim == C.ndim == D.ndim, "Invalid System Matrices dimensions"
+        self._A, self._B, self._C, self._D = A, B, C, D
+        self._c1, self._c2 = c1, c2
+
+    @property
+    def c1(self):
+        return self._c1
+    
+    @property
+    def c2(self):
+        return self._c2
+    
+    def forward(self, x, u):
+        r'''
+        Args:
+            x (:obj:`Tensor`): The state of LTI system,
+            u (:obj:`Tensor`): The input of LTI system.
+
+        Return:
+            Tensor: The solved batched tensor, which means state-transition and observation of the system.
```

Comment:
The state of the next timestamp (state-transition) and the system output (observation). 
Return type: should be "tuple of tensors". (you have both the state and observation)